### PR TITLE
network/tc: Avoid concurrent set modification in tclass_drop()/qdisc_drop()

### PR DESCRIPTION
**Pull request by daandemeyer**
With the current algorithm, we can end up removing entries from the qdisc/tclass sets while having multiple open iterators over the sets at various positions which leads to assertion failures in the hashmap logic as it's only safe to remove the "current" entry.

To avoid the problem, let's split up marking and dropping of tclasses and qdiscs. First, we recursively iterate tclasses/qdiscs and mark all that need to be removed. Next, we iterate once over tclasses and qdiscs and remove all marked entries.

### CODE REVIEW

File: src/network/tc/tclass.c [hunk 1]
```diff
@@ -252,37 +252,55 @@ static void log_tclass_debug(TClass *tclass, Link *link, const char *str) {
...
+                        log_tclass_debug(tclass, link, "Forgetting");
+                        tclass_free(tclass);
+                } else
+                        log_tclass_debug(tclass, link, "Removed");
         }
+}
 
-        tclass_unmark(tclass);
-        tclass_enter_removed(tclass);
+TClass *tclass_drop(TClass *tclass) {
```
Comment (yuwata) [inline src/network/tc/tclass.c]: Nitpick, but `TClass* tclass_drop()` rather than `TClass *tclass_drop()`, please.

File: src/network/tc/qdisc.c [hunk 2]
```diff
@@ -285,37 +285,55 @@ int link_find_qdisc(Link *link, uint32_t handle, const char *kind, QDisc **ret)
...
+        assert(qdisc->link);
 
-        if (qdisc->state == 0) {
-                log_qdisc_debug(qdisc, link, "Forgetting");
-                qdisc = qdisc_free(qdisc);
-        } else
-                log_qdisc_debug(qdisc, link, "Removed");
+        qdisc_mark_recursive(qdisc);
+        link_qdisc_drop_marked(qdisc->link);
+        link_tclass_drop_marked(qdisc->link);
```
Comment (yuwata) [inline src/network/tc/qdisc.c:335]: `qdisc` may be invalidated here. So, please take a reference of `qdisc->link` before calling `_drop_marked()`.

File: src/network/tc/tclass.c [hunk 3]
```diff
@@ -252,37 +252,55 @@ static void log_tclass_debug(TClass *tclass, Link *link, const char *str) {
...
+        assert(tclass->link);
 
-        if (tclass->state == 0) {
-                log_tclass_debug(tclass, link, "Forgetting");
-                tclass = tclass_free(tclass);
-        } else
-                log_tclass_debug(tclass, link, "Removed");
+        tclass_mark_recursive(tclass);
+        link_tclass_drop_marked(tclass->link);
+        link_qdisc_drop_marked(tclass->link);
```
Comment (yuwata) [inline src/network/tc/tclass.c:301]: Also here.

**Review by yuwata — CHANGES_REQUESTED** [review]
Thank you. The basic change looks good, but CIs are unhappy.

**Comment by yuwata** [comment]
Could you mention the previous offending commit 632d321050f58fe1b5bed7cfe769d212377c0301 in the commit message?

**Comment by yuwata** [comment]
The commit 632d321050f58fe1b5bed7cfe769d212377c0301 has been backported to stable repository. Hence, this also needs to be backported.

**Comment by yuwata** [comment]
> Addressed comments and added two commits required to run systemd-networkd-tests.py in mkosi, PTAL!

Nice! Thank you!

**Review by yuwata — APPROVED** [review]

File: test/meson.build [hunk 4]
```diff
@@ -29,7 +29,7 @@ if install_tests
                 # 'follow_symlinks' once we require Meson 1.3.0 or greater, see:
                 # https://github.com/mesonbuild/meson/commit/0af126fec798d6dbb0d1ad52168cc1f3f1758acd
                 if rsync.found()
-                        rsync_r = rsync.full_path() + ' -rlpt --exclude .gitattributes -- "@0@" "${DESTDIR:-}@1@"'
+                        rsync_r = rsync.full_path() + ' -rLpt --exclude .gitattributes -- "@0@" "${DESTDIR:-}@1@"'
```
Comment (bluca) [inline test/meson.build:32]: if I understand correctly the comment above, rsync was added specifically to copy symlinks, and the default in install-subdir is to follow them, so can this be dropped entirely then?

Comment (daandemeyer) [reply test/meson.build:32]: @bluca It depends on whether we still get the ugly warning or not

Comment (bluca) [reply test/meson.build:32]: ok, we can revisit it later

**Comment by keszybz** [comment]
> The commit [632d321](https://github.com/systemd/systemd/commit/632d321050f58fe1b5bed7cfe769d212377c0301) has been backported to stable repository. Hence, the last commit also needs to be backported.

Thanks. Done.

**Pull request merged**